File: Grammars/LUA/Calculator.vb
Imports NLua
Imports Utility
Imports Microsoft.VisualBasic.FileIO
Imports System.Xml

Namespace LUA
    Friend Class Calculator

        Private _calculator As Grammars.Calculator
        Private _grammar As BaseGrammar
        Private _lua As NLua.Lua

        Public Sub New(ByVal grammar As LUAGrammar,
                       ByVal lua As NLua.Lua)
            _grammar = grammar
            _lua = lua
            _calculator = New Grammars.Calculator(_grammar)
        End Sub

        Public Function ExpandPath(ByVal path As String) As String
            Dim value As String = Environment.ExpandEnvironmentVariables(path)
            If Not IO.Path.IsPathRooted(value) Then
                value = IO.Path.Combine(IO.Path.GetDirectoryName(_grammar.FilePath), value)
            End If
            Return value
        End Function

        Public Function Roll(ByVal ParamArray params() As Object) As Int32
            Return Grammars.Calculator.DiceRoll(params)
        End Function

        Public Function Rnd(ByVal min As Int32, ByVal max As Int32) As Int32
            Return BaseGrammar.Random.Next(min, max + 1)
        End Function

        Public Function TargetRoll(ByVal ParamArray params() As Object) As Int32
            Return Grammars.Calculator.TargetRoll(params)
        End Function

        Public Function Generate(ByVal grammarName As String,
                                 ByVal maxLength As Int32) As String
            Return _calculator.Generate(grammarName, maxLength)
        End Function

        Public Function Generate(ByVal grammarName As String,
                                 ByVal maxLength As Int32,
                                 ByVal parameters As LuaTable) As String
            Return _calculator.Generate(grammarName, maxLength, parameters)
        End Function

        Public Function Evaluate(ByVal expression As String) As Object
            Return _calculator.Evaluate(expression)
        End Function

        Public Function SelectFromTable(ByVal items As LuaTable) As Object
            Dim maxRoll As Int32 = 0
            Dim index As Int32 = 0

            For Each item As Int32 In items.Keys
                If maxRoll < item Then maxRoll = item
            Next

            index = BaseGrammar.Random.Next(1, maxRoll + 1)

            For Each item As Int32 In items.Keys
                If index <= item Then Return items.Item(item)
            Next

            Return Nothing
        End Function

        Public ReadOnly Property LastRollResult As Grammars.Calculator.DieRollResult
            Get
                Return Grammars.Calculator.LastRollResult
            End Get
        End Property

        Public Function FileToString(ByVal filePath As String) As String
            Dim path As String = filePath
            If Not IO.Path.IsPathRooted(path) Then path = IO.Path.Combine(IO.Path.GetDirectoryName(_grammar.FilePath), filePath)
            Return IO.File.ReadAllText(filePath)
        End Function

        Public Function FileToTable(ByVal filePath As String) As String()
            Return IO.File.ReadAllLines(filePath)
        End Function

        Public Function XMLFileToTable(ByVal filePath As String) As LuaTable
            Dim value As LuaTable
            Dim document As New XmlDocument
            document.Load(filePath)

            value = ProcessXMLNode(document.DocumentElement)

            Return value
        End Function

        Private Function ProcessXMLNode(ByVal parent As XmlNode) As LuaTable
            Dim value As LuaTable = CreateTable()
            Dim indexed As Boolean = False
            Dim index As Int32 = 1

            If parent.Attributes("indexed") IsNot Nothing Then
                indexed = CBool(parent.Attributes("indexed").Value)
            End If

            For Each child As XmlNode In parent.ChildNodes
                If child.HasChildNodes AndAlso child.FirstChild.NodeType = XmlNodeType.Element Then
                    If indexed Then
                        value(index) = ProcessXMLNode(child)
                        index += 1
                    Else
                        value(child.Name) = ProcessXMLNode(child)
                    End If
                ElseIf child.NodeType <> XmlNodeType.Comment Then
                    If indexed Then
                        value(index) = child.InnerText
                        index += 1
                    Else
                        value(child.Name) = child.InnerText
                    End If
                End If
            Next
            Return value
        End Function

        Private Function CreateTable() As LuaTable
            Return CType(_lua.DoString("return {}")(0), LuaTable)
        End Function

        Public Function CSVToTable(ByVal csvPath As String) As LuaTable
            Dim parser As New TextFieldParser(csvPath)
            Dim header() As String
            Dim value As LuaTable = CreateTable()
            Dim index As Int32 = 1
            parser.TextFieldType = FieldType.Delimited
            parser.HasFieldsEnclosedInQuotes = True
            parser.SetDelimiters(",")

            header = parser.ReadFields

            Do While Not parser.EndOfData
                Dim line() As String = parser.ReadFields
                Dim record As LuaTable = CreateTable()
                For i As Int32 = 0 To line.Length - 1
                    record(header(i)) = line(i)
                Next
                value(index) = record
                index += 1
            Loop

            Return value
        End Function
    End Class
End Namespace

File: Grammars/LUA/Calculator.cs
using System;
using System.IO;
using System.Linq;
using System.Xml;
using Microsoft.VisualBasic.FileIO;
using NLua;

namespace Grammars.LUA
{
    internal class Calculator
    {
        private readonly Grammars.Calculator _calculator;
        private readonly BaseGrammar _grammar;
        private readonly Lua _lua;

        public Calculator(LUAGrammar grammar, Lua lua)
        {
            _grammar = grammar;
            _lua = lua;
            _calculator = new Grammars.Calculator(_grammar);
        }

        public Grammars.Calculator.DieRollResult LastRollResult
        {
            get { return Grammars.Calculator.LastRollResult; }
        }

        public string ExpandPath(string path)
        {
            var value = Environment.ExpandEnvironmentVariables(path);
            if (!Path.IsPathRooted(value))
            {
                value = Path.Combine(Path.GetDirectoryName(_grammar.FilePath), value);
            }
            return value;
        }

        public int Roll(params object[] @params)
        {
            return Grammars.Calculator.DiceRoll(@params);
        }

        public int Rnd(int min, int max)
        {
            return BaseGrammar.Random.Next(min, max + 1);
        }

        public int TargetRoll(params object[] @params)
        {
            return Grammars.Calculator.TargetRoll(@params);
        }

        public string Generate(string grammarName, int maxLength)
        {
            return _calculator.Generate(grammarName, maxLength);
        }

        public string Generate(string grammarName, int maxLength, LuaTable parameters)
        {
            return _calculator.Generate(grammarName, maxLength, parameters);
        }

        public dynamic Evaluate(string expression)
        {
            return _calculator.Evaluate(expression);
        }

        public dynamic SelectFromTable(LuaTable items)
        {
            var maxRoll = items.Keys
                               .Cast<int>()
                               .Concat(new[] {0})
                               .Max();

            var index = BaseGrammar.Random.Next(1, maxRoll + 1);
            foreach (int item in items.Keys)
            {
                if (index <= item)
                {
                    return items[Convert.ToString(item)];
                }
            }

            return null;
        }

        public string FileToString(string filePath)
        {
            var path = filePath;
            if (!Path.IsPathRooted(path))
            {
                path = Path.Combine(Path.GetDirectoryName(_grammar.FilePath), filePath);
            }
            return File.ReadAllText(filePath);
        }

        public string[] FileToTable(string filePath)
        {
            return File.ReadAllLines(filePath);
        }

        public LuaTable XMLFileToTable(string filePath)
        {
            var document = new XmlDocument();
            document.Load(filePath);

            var value = ProcessXMLNode(document.DocumentElement);

            return value;
        }

        private LuaTable ProcessXMLNode(XmlNode parent)
        {
            var value = CreateTable();
            var indexed = false;
            var index = 1;

            if (parent.Attributes["indexed"] != null)
            {
                indexed = bool.Parse(parent.Attributes["indexed"].Value);
            }

            foreach (XmlNode child in parent.ChildNodes)
            {
                if (child.HasChildNodes && child.FirstChild.NodeType == XmlNodeType.Element)
                {
                    if (indexed)
                    {
                        value[index] = ProcessXMLNode(child);
                        index++;
                    }
                    else
                    {
                        value[child.Name] = ProcessXMLNode(child);
                    }
                }
                else if (child.NodeType != XmlNodeType.Comment)
                {
                    if (indexed)
                    {
                        value[index] = child.InnerText;
                        index++;
                    }
                    else
                    {
                        value[child.Name] = child.InnerText;
                    }
                }
            }
            return value;
        }

        private LuaTable CreateTable()
        {
            return ((LuaTable) (_lua.DoString("return {}")[0]));
        }

        public LuaTable CSVToTable(string csvPath)
        {
            var parser = new TextFieldParser(csvPath)
            {
                TextFieldType = FieldType.Delimited,
                HasFieldsEnclosedInQuotes = true
            };
            parser.SetDelimiters(",");

            var header = parser.ReadFields();

            var value = CreateTable();
            for (var index = 1; !parser.EndOfData; index++)
            {
                var line = parser.ReadFields();
                var record = CreateTable();
                for (var i = 0; i <= line.Length - 1; i++)
                {
                    record[header[i]] = line[i];
                }
                value[index] = record;
            }

            return value;
        }
    }
}